annotations/canonicalizer.go

```diff
@@ -0,0 +1,68 @@
+package annotations
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"sort"
+	"strings"
+)
+
+type Canonicalizer struct {
+	sorterFactory func(ann []Annotation) sort.Interface
+}
+
+type annotationSorter struct {
+	ann []Annotation
+}
+
+func NewCanonicalizer(sorterFactory func(ann []Annotation) sort.Interface) *Canonicalizer {
+	return &Canonicalizer{sorterFactory}
+}
+
+func (c *Canonicalizer) canonicalize(in []Annotation) []Annotation {
+	out := make([]Annotation, len(in))
+	for i, ann := range in {
+		out[i] = *c.deplete(ann)
+	}
+
+	sort.Sort(c.sorterFactory(out))
+	return out
+}
+
+func (c *Canonicalizer) deplete(in Annotation) *Annotation {
+	return &Annotation{Predicate: in.Predicate, ConceptId: in.ConceptId}
+}
+
+// Hash hashes the given payload in SHA224 + Hex
+func (c *Canonicalizer) hash(ann []Annotation) string {
```

Comment:
This seems to be used in the test only. Is it also here in anticipation of the next Hash Wars story?

Reply:
Yes, for now it functions only as a demonstration/test of canonicalization giving reproducible hash values. (It's lifted from https://github.com/Financial-Times/nativerw/blob/master/resources/hashing.go#L17)